### PR TITLE
Added autograd support for torch.orgqr

### LINKED ISSUE #50104: Support CUDA and backpropagation in `torch.orgqr`
## 🚀 Feature

As of PyTorch 1.7, `torch.orgqr` only supports CPU tensors and does not implement gradients. 
Feature proposal: add support for CUDA tensors and gradients wrt inputs.

## Motivation

This function is required to perform orthogonal parameterization of matrices (not to be confused with orthogonal initialization). Given an input matrix of shape `d x r` with `d >= r`, `orgqr` produces an orthogonal matrix of the same shape `d x r`. The output spans the manifold of Stiefel frames of size `d x r`, and hence, provides precise orthogonal constraints by construction. This can be used further, e.g. as a weight matrix in an RNN transition layer or any other place requiring strict orthogonality. More motivation can be found here and references therein: https://github.com/pytorch/pytorch/issues/42243
Without GPU, the function is slow. Without gradients, parameterization becomes impossible. Once the requested feature is implemented, and once this PR https://github.com/pytorch/pytorch/pull/33344 is merged, it will become possible to perform precise orthogonal parameterization of weight matrices in Linear, all flavors of Convolutional, and Embedding layers.

## Pitch

`torch.orgqr` to support the two requested properties.
As hinted by @Lezcano, MAGMA already implements a GPU side of the function: https://www.icl.utk.edu/~mgates3/docs/magma-proto.html
The gradients can be taken from my package https://github.com/toshas/torch-householder/blob/master/torch_householder/householder.cpp#L58

## Alternatives

The code from my package https://github.com/toshas/torch-householder implements an efficient drop-in replacement for `torch.orgqr` satisfying the requested properties, and I don't mind donating this code. The forward pass might be slower than that of MAGMA -- no comparison was done.

## Additional context

None so far

cc @ngimel @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @walterddr @IvanYashchuk

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This PR adds autograd support for `torch.orgqr`.

Since `torch.orgqr` is one of few functions that expose LAPACK's naming and all other linear algebra routines were renamed a long time ago, I also added a new function with a new name and `torch.orgqr` now is an alias for it.

The new proposed name is `householder_product`. For a matrix `input` and a vector `tau` LAPACK's orgqr operation takes columns of `input` (called Householder vectors or elementary reflectors) scalars of `tau` that together represent Householder matrices and then the product of these matrices is computed. See https://www.netlib.org/lapack/lug/node128.html.
Other linear algebra libraries that I'm aware of do not expose this LAPACK function, so there is some freedom in naming it. It is usually used internally only for QR decomposition, but can be useful for deep learning tasks now when it supports differentiation.

Resolves #50104

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
Probably complex gradgradcheck should be marked as slow.
```
============================================================ slowest 10 durations =============================================================
11.94s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_householder_product_cpu_complex128
3.36s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_householder_product_cpu_float64
0.96s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_householder_product_cuda_complex128
0.36s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_householder_product_cpu_complex128
0.15s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_householder_product_cpu_float64
0.13s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_householder_product_cpu_complex64
0.13s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_householder_product_cpu_complex128
0.12s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_householder_product_cpu_float32
0.12s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_householder_product_cpu_float64
0.09s call     test/test_ops.py::TestOpInfoCUDA::test_unsupported_dtypes_householder_product_cuda_int32
```

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -1290,15 +1297,24 @@ Tensor& orgqr_out_info(const Tensor& input, const Tensor& tau, Tensor& result, T
...
   return result;
 }
 
-Tensor& orgqr_out(const Tensor& input, const Tensor& tau, Tensor& result) {
+Tensor& householder_product_out(const Tensor& input, const Tensor& tau, Tensor& result) {
   TORCH_CHECK(input.dim() >= 2, "orgqr: input must have at least 2 dimensions.");
   TORCH_CHECK(input.size(-2) >= input.size(-1), "orgqr: input.shape[-2] must be greater than or equal to input.shape[-1]");
   TORCH_CHECK(input.size(-1) >= tau.size(-1), "orgqr: input.shape[-1] must be greater than or equal to tau.shape[-1]");
 
+  TORCH_CHECK(input.dim() - tau.dim() == 1, "orgqr: Expected tau to have one dimension less than input, but got tau.ndim equal to ",
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1309]: Wouldn't the names for these operations start with "householder_product:" now? Or maybe even "torch.linalg.householder_product():"?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -1351,12 +1367,20 @@ Tensor& orgqr_out(const Tensor& input, const Tensor& tau, Tensor& result) {
...
 }
 
-Tensor orgqr(const Tensor& input, const Tensor& tau) {
+Tensor householder_product(const Tensor& input, const Tensor& tau) {
   Tensor result = at::empty({0}, input.options());
   result = at::orgqr_outf(input, tau, result);
   return result;
 }
 
+Tensor& orgqr_out(const Tensor& input, const Tensor& tau, Tensor& result) {
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1376]: Add a comment her explaining this alias

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -6000,14 +6000,20 @@
...
-    CPU: orgqr_out
+    CPU: householder_product_out
 
+# orgqr, alias for householder_product
 - func: orgqr(Tensor self, Tensor input2) -> Tensor
   variants: method, function
-  dispatch:
-    CPU: orgqr
+
+- func: orgqr.out(Tensor self, Tensor input2, *, Tensor(a!) out) -> Tensor(a!)
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:6016]: So orgqr was always CPU-only and this is also CPU-only for now?

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -6628,21 +6686,18 @@ def merge_dicts(*dicts):
...
 
-Computes the orthogonal matrix `Q` of a QR factorization, from the `(input, input2)`
+Computes the orthogonal matrix `Q` of a QR factorization, from the `(input, tau)`
 tuple returned by :func:`torch.geqrf`.
 
-This directly calls the underlying LAPACK function `?orgqr`.
-See `LAPACK documentation for orgqr`_ for further details.
+This directly calls the underlying LAPACK function `orgqr`.
+
+Alias for :func:`torch.householder_product`.
```
Comment (mruberry) [inline torch/_torch_docs.py:6696]: Maybe we should just have the "Alias for ..." line? The rest is redundant with it.

**Comment by mruberry** [comment]
ASAN failure looks real.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 5]
```diff
@@ -2987,6 +2987,105 @@ bool any_variable_defined(variable_list& variables) {
   return false;
 }
 
+std::tuple<Tensor, Tensor> householder_product_backward(const Tensor& grad, const Tensor& input, const Tensor& tau) {
+  if (!grad.defined()) {
+    return std::tuple<Tensor, Tensor>(Tensor(), Tensor());
+  }
+
+  // LAPACK's orgqr operation computes a product of Householder matrices.
```
Comment (mruberry) [inline torch/csrc/autograd/FunctionsManual.cpp:2995]: Excellent work on the comments here.

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -1005,6 +1005,35 @@ def sample_inputs(self, device, dtype, requires_grad=False):
...
+    """
+    This function generates input for torch.householder_product (torch.orgqr).
+    The first argument should be a square matrix or batch of square matrices, the second argument is a vector or batch of vectors.
+    Empty, square, rectangular, batched square and batched rectangular input is generated.
+    """
+    samples = (
+        SampleInput((make_tensor((0, 0), device, dtype, low=None, high=None, requires_grad=requires_grad),
+                    make_tensor((0,), device, dtype, low=None, high=None, requires_grad=requires_grad))),
+
+        SampleInput((make_tensor((S, S), device, dtype, low=-2, high=2, requires_grad=requires_grad),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1018]: Would you elaborate on the range here?

File: torch/_torch_docs.py [hunk 7]
```diff
@@ -3672,6 +3672,64 @@ def merge_dicts(*dicts):
...
+or columns of each matrix in a batched :attr:`input`.
+:attr:`tau` is a tensor of scalar scale factors for each Householder vector.
+For the single matrix :attr:`input` taking its `i`-th column as `vᵢ` and `τ = tau[i]`
+gives the `i`-th Householder matrix as `Hᵢ = I − τ vᵢ vᵢᴴ`.
+The result of this function is `H₁ H₂ ... Hᵣ`, where `r` is equal to `tau.shape[-1]`.
+This function is commonly used together with :func:`torch.geqrf`
+to explitly form the `Q` matrix of the QR decomposition.
+See `Representation of Orthogonal or Unitary Matrices`_ for further details.
+
+.. note:: LAPACK's `orgqr` is used for the computations.
```
Comment (mruberry) [inline torch/_torch_docs.py:3691]: Let's be explicit this only supports CPU inputs.

File: torch/_torch_docs.py [hunk 8]
```diff
@@ -3672,6 +3672,64 @@ def merge_dicts(*dicts):
...
+householder_product(input, tau, *, out=None) -> Tensor
+
+Computes the product of Householder matrices.
+
+Householder matrices are stored in compressed vector form as columns of a `(m × n)` matrix :attr:`input`,
+or columns of each matrix in a batched :attr:`input`.
+:attr:`tau` is a tensor of scalar scale factors for each Householder vector.
+For the single matrix :attr:`input` taking its `i`-th column as `vᵢ` and `τ = tau[i]`
+gives the `i`-th Householder matrix as `Hᵢ = I − τ vᵢ vᵢᴴ`.
+The result of this function is `H₁ H₂ ... Hᵣ`, where `r` is equal to `tau.shape[-1]`.
```
Comment (mruberry) [inline torch/_torch_docs.py:3686]: New paragraph here

File: torch/_torch_docs.py [hunk 9]
```diff
@@ -3672,6 +3672,64 @@ def merge_dicts(*dicts):
...
+add_docstr(torch.householder_product,
+           r"""
+householder_product(input, tau, *, out=None) -> Tensor
+
+Computes the product of Householder matrices.
+
+Householder matrices are stored in compressed vector form as columns of a `(m × n)` matrix :attr:`input`,
+or columns of each matrix in a batched :attr:`input`.
+:attr:`tau` is a tensor of scalar scale factors for each Householder vector.
+For the single matrix :attr:`input` taking its `i`-th column as `vᵢ` and `τ = tau[i]`
```
Comment (mruberry) [inline torch/_torch_docs.py:3684]: What should the shape of tau be?

File: torch/_torch_docs.py [hunk 10]
```diff
@@ -3672,6 +3672,64 @@ def merge_dicts(*dicts):
     tensor([ 0.,  2.,  1.,  0.])
 """.format(**common_args))
 
+add_docstr(torch.householder_product,
+           r"""
+householder_product(input, tau, *, out=None) -> Tensor
```
Comment (mruberry) [inline torch/_torch_docs.py:3677]: Should this be torch.linalg.householder_product()?

**Review by mruberry — COMMENTED** [review]
Hey @IvanYashchuk! Overall this PR looks very good and I'm sorry you've been waiting for this review.

I made a few inline comments. My biggest question is: should this be in the torch.linalg namespace? 

Also it looks like one of the test failures is related to the PR. Looking forward to an update here!

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/native_functions.yaml:6016]: Yes, it was always CPU-only. CUDA support is in https://github.com/pytorch/pytorch/pull/51348 which now passes all CI checks and can be merged. After https://github.com/pytorch/pytorch/pull/51348 is merged I'll update this PR.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:1309]: Yes, I should change that.

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (IvanYashchuk) [reply torch/testing/_internal/common_methods_invocations.py:1018]: Each column of the matrix is getting multiplied many times leading to very large values for the Jacobian matrix entries and making the finite-difference result of grad check less accurate. That's why gradcheck with the default range [-9, 9] fails.

**Comment by IvanYashchuk** [comment]
> should this be in the torch.linalg namespace?

Currently `torch.linalg` contains only `numpy.linalg` compatible functions. I think it was not decided yet whether all linear algebra related functions (the ones that take matrices or batch of matrices as input) should live under `torch.linalg`. If we decide that it should be that way then we can immediately put this function under linalg namespace. It should be consistent and all other functions should be moved then as well for 1.9 release possibly with SciPy compatibility.

**Comment by mruberry** [comment]
> > should this be in the torch.linalg namespace?
> 
> Currently `torch.linalg` contains only `numpy.linalg` compatible functions. I think it was not decided yet whether all linear algebra related functions (the ones that take matrices or batch of matrices as input) should live under `torch.linalg`. If we decide that it should be that way then we can immediately put this function under linalg namespace. It should be consistent and all other functions should be moved then as well for 1.9 release possibly with SciPy compatibility.

I think you're part of that decision as a torch.linalg lead ;)

Let me add some context to help inform our decision:

- A frequent complaint I hear is that numpy.linalg vs. numpy makes no sense. So it'd be nice if we had some principles determining when something should be in torch.linalg vs torch. Conceptually it'd be nice if all "linear algebra" operations were just in torch.linalg. Where this runs into some trouble is that very common operations, like matmuls, might be difficult to actually deprecate and move, but maybe a good first rule would be "all new linear algebra operations go in torch.linalg."
- @kurtamohler already helped us design torch.linalg.vector_norm and torch.linalg.matrix_norm, so the namespace will already contain operations not in NumPy or SciPy's linalg namespaces. I think this is a cool feature we'll want to talk about in our 1.9 blog post (that is, how we're going beyond in our design of the linalg module).
- We'll also probably have new features in our linalg module, like those coming from the historic PyTorch linalg functions and  @Balandat's idea for being able to skip error syncs on CUDA devices.

We don't have to decide this immediately. In fact we should probably schedule another torch.linalg sync. We could talk about it for a bit then, too, if you like.

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:1018]: Cool; would you add a comment explaining that?

**Comment by nikitaved** [comment]
@IvanYashchuk , sorry for the delay. I will check the math soon!

File: torch/_torch_docs.py (hunk 7 above)
Comment (IvanYashchuk) [reply torch/_torch_docs.py]: Now it supports GPU as well.

**Comment by IvanYashchuk** [comment]
@mruberry I moved `householder_product` to the `linalg` module.
I also updated this PR so that CUDA is also supported with https://github.com/pytorch/pytorch/pull/51348. Gradchecks tests revealed that cuSOLVER doesn't return the correct result for the empty `tau` argument, which is fixed in https://github.com/pytorch/pytorch/pull/52637/commits/3a99d1f3c7d6643663e1b00f45290811c26e2f40.

**Comment by mruberry** [comment]
The ASAN failure looks real, unfortunately.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 11]
```diff
@@ -3097,6 +3097,105 @@ bool any_variable_defined(variable_list& variables) {
...
+  auto start_j = tau.size(-1) - 1;
+  for (int64_t j = start_j; j >= 0; j--) {
+    auto v = input_.index({"...", Slice(), j});
+    auto v1 = v, v2 = v;
+
+    // we need to recompute input[j] * at::outer(v, v)
+    auto tau_unsqueezed = tau.index({"...", j}).unsqueeze(-1);  // tau[..., j][:, None]
+    auto tau_unsqueezed2 = tau_unsqueezed.unsqueeze(-1);  // tau[..., j][:, None, None]
+    auto v3 = at::matmul(v1.unsqueeze(-1), v2.conj().unsqueeze(-2));  // batch outer product, at::outer doesn't work for batched input
+    auto v4 = tau_unsqueezed2 * v3;
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:3164]: `v3` can be computed as ` v.unsqueeze(-1) * v`.

Comment (IvanYashchuk) [reply torch/csrc/autograd/FunctionsManual.cpp:3164]: Yes, `v.unsqueeze(-1) * v.conj().unsqueeze(-2)` will work.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 12]
```diff
@@ -3097,6 +3097,105 @@ bool any_variable_defined(variable_list& variables) {
...
+  auto d_input = at::zeros_like(input);
+  auto d_tau = at::zeros_like(tau);
+
+  auto d_result = at::zeros(square_shape, grad.options());
+  d_result.index_put_({"...", Slice(), Slice(None, input.size(-1))}, grad); // d_result[..., :, :input.shape[-1]] = grad
+
+  auto start_j = tau.size(-1) - 1;
+  for (int64_t j = start_j; j >= 0; j--) {
+    auto v = input_.index({"...", Slice(), j});
+    auto v1 = v, v2 = v;
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:3158]: I believe you can replace indexing with ellipsis with `narrow`, should be faster.

Comment (IvanYashchuk) [reply torch/csrc/autograd/FunctionsManual.cpp:3158]: In my tests, replacing `.index` here and later in code with `.narrow` doesn't make any difference.

**Comment by nikitaved** [comment]
@mruberry , in our discussion with @IvanYashchuk  we came to the conclusion that his PR could be merged as of now.
However, there an efficient algorithm for this backward which is simpler, but requires some theoretical insight and basic properties of reflectors. The cool part of this algorithm is that it requires only a single matrix-matrix product, the rest could be done with matrix-vector routines, which results in total complexity of 2-4 matrix products. I could submit a PR later.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 12 above)
Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:3158]: Of course, compared to matrix products it is almost nothing, so, I guess, it is primarily a taste thingy. I have just never seen `index` used in ATen, so good to know this is yet another tool available :)

**Comment by IvanYashchuk** [comment]
@mruberry ASAN failure is fixed. CI failures are not related. It's ready for another round of review and probably can be merged.

File: torch/linalg/__init__.py [hunk 13]
```diff
@@ -367,6 +367,64 @@
...
+:attr:`tau` is a tensor of scalar scale factors for each Householder vector.
+For the single matrix :attr:`input` taking its `i`-th column as `vᵢ` and `τ = tau[i]`
+gives the `i`-th Householder matrix as `Hᵢ = I − τ vᵢ vᵢᴴ`.
+
+The result of this function is `H₁ H₂ ... Hᵣ`, where `r` is equal to `tau.shape[-1]`.
+This function is commonly used together with :func:`torch.geqrf`
+to explitly form the `Q` matrix of the QR decomposition.
+See `Representation of Orthogonal or Unitary Matrices`_ for further details.
+
+.. note:: LAPACK's `orgqr` is used for the computations.
```
Comment (mruberry) [inline torch/linalg/__init__.py:386]: What about on CUDA?

**Review by mruberry — APPROVED** [review]
Thank you for helping review this @nikitaved and providing an update. And great work to both you and @IvanYashchuk!

It's exciting to have new linalg functions in torch.linalg!

This looks good but you would rebase past the doc failures? I'm also curious what's going on the XLA build. I made one tiny docs note about explaining what's called on CUDA (since orgqr now supports CUDA) that we might as well resolve during the rebase.

File: torch/linalg/__init__.py (hunk 13 above)
Comment (IvanYashchuk) [reply torch/linalg/__init__.py:386]: Added a note: "For CUDA inputs, cuSOLVER's `orgqr` routine is used if CUDA version >= 10.1.243."

**Comment by IvanYashchuk** [comment]
@mruberry I updated the branch and added a note that cuSOLVER is used for CUDA inputs. Let's see whether doc and xla CI failures go away.

**Pull request closed without merging**